package/gluon-mesh-vpn-wireguard/src/key-translate.c

```diff
@@ -0,0 +1,69 @@
+#include <libubox/utils.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define FASTD_KEY_LEN_HEX 64
+
+static inline void invalid_key(void) {
+	fprintf(stderr, "Error: Invalid fastd key provided.\n");
+	exit(1);
+}
+
+static inline void usage(void) {
+	fprintf(stderr, "Usage: key-translate [-n] <fastd secret>\n");
+	fprintf(stderr, "-n     do not output the trailing newline\n");
+	exit(1);
+}
+
+static inline bool read_key(uint8_t key[32], const char *hexkey) {
+        if ((strlen(hexkey) != 64) || (strspn(hexkey, "0123456789abcdefABCDEF") != 64))
+                return false;
+
+        size_t i;
+        for (i = 0; i < 32; i++)
+                sscanf(&hexkey[2 * i], "%02hhx", &key[i]);
+
+        return true;
+}
+
+int main(int argc, char *argv[]) {
+	bool newline = true;
+	if (argc < 2 || argc > 3) {
+		usage();
+	} else if (argc==3) {
+		if (strcmp(argv[1], "-n")) {
+			usage();
+		} else {
+			newline=false;
+		}
+	}
+	int fastd_pos = argc-1;
+	int fastd_len = strlen(argv[fastd_pos]);
+
+	if (FASTD_KEY_LEN_HEX != fastd_len)
+		invalid_key();
+
+	uint8_t fastd_keystore[32];
+	read_key(fastd_keystore, argv[fastd_pos]);
+
+	// l is always 45 ((((32) + 2) / 3) * 4 + 1)
+	int l = B64_ENCODE_LEN(32);
+	char *str = malloc(l);
```

Comment:
Check if allocation was succesful before using buffer

Reply:
resolved in 65fd9de1c861d7661f31c42ea4051ea7b66ced89